kong/clustering/control_plane.lua

```diff
@@ -562,8 +571,9 @@ local function push_config_loop(premature, self, push_config_semaphore, delay)
       else
         ngx_log(ngx_ERR, _log_prefix, "export and pushing config failed: ", err)
       end
+    end
 
-    elseif err ~= "timeout" then
+    if err ~= "timeout" then
```

Comment:
if `ok` is truth-y but `self.clients` is empty, this branch will be followed because `err` will be `nil` and generate a confusing error log entry (I think).

Maybe it would be more straightforward to put the `self.clients` check into `push_config`?

```lua
function _M:push_config()
  if not next(self.clients) then
    return true
  end

  local start = ngx_now()

  local payload, err = self:export_deflated_reconfigure_payload()
  if not payload then
  -- ...
end
```

Reply:
Alternatively, maybe the `self.clients` check could be pushed to the event handler? This would be simple for both `control_plane` and `wrpc_control_plane`:

```lua
  kong.worker_events.register(function(_)
    if next(self.clients) and push_config_semaphore:count() <= 0 then
      -- the following line always executes immediately after the `if` check
      -- because `:count` will never yield, end result is that the semaphore
      -- count is guaranteed to not exceed 1
      push_config_semaphore:post()
    end
  end, "clustering", "push_config")
```

Reply:
@flrgh The first approach has some issues, if we `return true` inside `push_config`, then the timer of `push_config_loop` is still reset which means that loop will wait for `db_update_frequency` before even pushing out another updated config.

The second issue might work (I did not spot immediate issue with it), but it seems to be a bit more dangerous as there are a lot of asynchronous going on between worker events and the semaphore. It's a bit harder to argue the correctness.

My thought:

If we change the `push_config_loop` function to do the following:

```
while not exiting() do
    local ok, err = push_config_semaphore:wait(1)
    if exiting() then
      return
    end

    if ok then
      if next(self.clients) do
        ok, err = pcall(self.push_config, self)
        ...
```

Does this also work? It seems to be simpler than the current approach.